File: java8/src/main/java/folio/codinginterview/infrastructure/server/DummyServer.java
package folio.codinginterview.infrastructure.server;

import folio.codinginterview.application.usecase.asset.GetAssetUsecase;
import folio.codinginterview.application.usecase.order.AdditionalBuyOrderUsecase;
import folio.codinginterview.application.usecase.order.NewOrderUsecase;
import folio.codinginterview.application.usecase.order.RebalanceOrderUsecase;
import folio.codinginterview.application.usecase.portfolio.GetLatestPortfolioUsecase;
import folio.codinginterview.application.usecase.portfolio.UpdatePortfolioUsecase;
import folio.codinginterview.infrastructure.repository.AccountRepositoryImpl;
import folio.codinginterview.infrastructure.repository.PortfolioRepositoryImpl;
import folio.codinginterview.presentation.AssetController;
import folio.codinginterview.presentation.OrderController;
import folio.codinginterview.presentation.PortfolioController;

public final class DummyServer {
    private final AssetController assetController;
    private final PortfolioController portfolioController;
    private final OrderController orderController;

    public DummyServer(
            AssetController assetController,
            PortfolioController portfolioController,
            OrderController orderController
    ) {
        this.assetController = assetController;
        this.portfolioController = portfolioController;
        this.orderController = orderController;
    }

    public AssetController assetController() { return assetController; }

    public PortfolioController portfolioController() { return portfolioController; }

    public OrderController orderController() { return orderController; }

    public static DummyServer defaultInstance() {
        PortfolioRepositoryImpl portfolioRepository = new PortfolioRepositoryImpl();
        AccountRepositoryImpl accountRepository = new AccountRepositoryImpl();

        GetAssetUsecase getAssetUsecase = new GetAssetUsecase(accountRepository);
        GetLatestPortfolioUsecase getLatestPortfolioUsecase = new GetLatestPortfolioUsecase(portfolioRepository);
        UpdatePortfolioUsecase updatePortfolioUsecase = new UpdatePortfolioUsecase(portfolioRepository);
        NewOrderUsecase newOrderUsecase = new NewOrderUsecase(accountRepository, portfolioRepository);
        AdditionalBuyOrderUsecase additionalBuyOrderUsecase = new AdditionalBuyOrderUsecase(accountRepository, portfolioRepository);
        RebalanceOrderUsecase rebalanceOrderUsecase = new RebalanceOrderUsecase(accountRepository, portfolioRepository);

        AssetController assetController = new AssetController(getAssetUsecase);
        PortfolioController portfolioController = new PortfolioController(getLatestPortfolioUsecase, updatePortfolioUsecase);
        OrderController orderController = new OrderController(newOrderUsecase, additionalBuyOrderUsecase, rebalanceOrderUsecase);

        return new DummyServer(assetController, portfolioController, orderController);
    }
}

File: php/src/CodingInterview/Infrastructure/Server/DummyServer.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Infrastructure\Server;

use Folio\CodingInterview\Application\Usecase\Asset\GetAssetUsecase;
use Folio\CodingInterview\Application\Usecase\Order\AdditionalBuyOrderUsecase;
use Folio\CodingInterview\Application\Usecase\Order\NewOrderUsecase;
use Folio\CodingInterview\Application\Usecase\Order\RebalanceOrderUsecase;
use Folio\CodingInterview\Application\Usecase\Portfolio\GetLatestPortfolioUsecase;
use Folio\CodingInterview\Application\Usecase\Portfolio\UpdatePortfolioUsecase;
use Folio\CodingInterview\Infrastructure\Repository\AccountRepositoryImpl;
use Folio\CodingInterview\Infrastructure\Repository\PortfolioRepositoryImpl;
use Folio\CodingInterview\Presentation\AssetController;
use Folio\CodingInterview\Presentation\OrderController;
use Folio\CodingInterview\Presentation\PortfolioController;

final class DummyServer
{
    public function __construct(
        public readonly AssetController $assetController,
        public readonly PortfolioController $portfolioController,
        public readonly OrderController $orderController,
    ) {}

    public static function default(): DummyServer
    {
        $portfolioRepository = new PortfolioRepositoryImpl();
        $accountRepository = new AccountRepositoryImpl();

        $getAssetUsecase = new GetAssetUsecase($accountRepository);
        $getLatestPortfolioUsecase = new GetLatestPortfolioUsecase($portfolioRepository);
        $updatePortfolioUsecase = new UpdatePortfolioUsecase($portfolioRepository);
        $newOrderUsecase = new NewOrderUsecase($accountRepository, $portfolioRepository);
        $additionalBuyOrderUsecase = new AdditionalBuyOrderUsecase($accountRepository, $portfolioRepository);
        $rebalanceOrderUsecase = new RebalanceOrderUsecase($accountRepository, $portfolioRepository);

        return new DummyServer(
            new AssetController($getAssetUsecase),
            new PortfolioController($getLatestPortfolioUsecase, $updatePortfolioUsecase),
            new OrderController($newOrderUsecase, $additionalBuyOrderUsecase, $rebalanceOrderUsecase),
        );
    }
}
